ext/js/language/language-transformer.js

```diff
@@ -120,18 +121,25 @@ export class LanguageTransformer {
         for (let i = 0; i < results.length; ++i) {
             const {text, conditions, trace} = results[i];
             for (const transform of this._transforms) {
-                if (!transform.suffixHeuristic.test(text)) { continue; }
+                if (!transform.heuristic.test(text)) { continue; }
 
                 const {name, rules} = transform;
                 for (let j = 0, jj = rules.length; j < jj; ++j) {
                     const rule = rules[j];
                     if (!LanguageTransformer.conditionsMatch(conditions, rule.conditionsIn)) { continue; }
-                    const {suffixIn, suffixOut} = rule;
-                    if (!text.endsWith(suffixIn) || (text.length - suffixIn.length + suffixOut.length) <= 0) { continue; }
+                    const {isInflected, deinflect} = rule;
+                    if (!isInflected.test(text)) { continue; }
+
+                    const isCycle = trace.some((frame) => frame.transform === name && frame.ruleIndex === j && frame.text === text);
+                    if (isCycle) {
+                        log.warn(new Error(`Cycle detected in transform[${name}] rule[${j}] for text: ${text}`));
```

Comment:
I think we can afford the performance required to check for cycles in transformer now, and it is now necessary since not all cycles can be prevented. Is there a way to warn here without interrupting?